gateway/api/views/files.py

```diff
@@ -276,48 +288,152 @@ def provider_list(self, request):
+        with tracer.start_as_current_span(
+            "gateway.files.provider_download", context=ctx
+        ):
+            username = request.user.username
+            requested_file_name = sanitize_file_name(
+                request.query_params.get("file", None)
+            )
+            provider_name = sanitize_name(request.query_params.get("provider", None))
+            function_title = sanitize_name(request.query_params.get("function", None))
+            working_dir = WorkingDir.PROVIDER_STORAGE
+
+            if (
+                requested_file_name is None
+                or function_title is None
+                or provider_name is None
+            ):
+                return Response(
+                    {
+                        "message": "File name, Qiskit Function title and Provider name are mandatory"  # pylint: disable=line-too-long
+                    },
+                    status=status.HTTP_400_BAD_REQUEST,
+                )
+
+            if not FileStorage.file_extension_is_valid(requested_file_name):
+                extensions = ", ".join(SUPPORTED_FILE_EXTENSIONS)
+                return Response(
+                    {
+                        "message": f"File name needs to have a valid extension: {extensions}"
+                    },
+                    status=status.HTTP_400_BAD_REQUEST,
+                )
+
+            if not self.user_has_provider_access(request.user, provider_name):
+                return Response(
+                    {"message": f"Provider {provider_name} doesn't exist."},
+                    status=status.HTTP_404_NOT_FOUND,
+                )
+
+            function = self.get_function(
+                user=request.user,
+                function_title=function_title,
+                provider_name=provider_name,
+            )
+            if not function:
+                return Response(
+                    {
+                        "message": f"Qiskit Function {provider_name}/{function_title} doesn't exist."  # pylint: disable=line-too-long
+                    },
+                    status=status.HTTP_404_NOT_FOUND,
+                )
+
+            file_storage = FileStorage(
+                username=username,
+                working_dir=working_dir,
+                function_title=function_title,
+                provider_name=provider_name,
+            )
+            result = file_storage.get_file(file_name=requested_file_name)
+            if result is None:
+                return Response(
+                    {"message": "Requested file was not found."},
+                    status=status.HTTP_404_NOT_FOUND,
+                )
+
+            file_wrapper, file_type, file_size = result
+            response = StreamingHttpResponse(file_wrapper, content_type=file_type)
+            response["Content-Length"] = file_size
+            response[
+                "Content-Disposition"
+            ] = f"attachment; filename={requested_file_name}"
```

Comment:
```suggestion
            if not self.user_has_provider_access(request.user, provider_name):
                return Response(
                    {"message": f"Provider {provider_name} doesn't exist."},
                    status=status.HTTP_404_NOT_FOUND,
                )

            return _download_file_response(
                request.user, 
                function_title, 
                provider_name, 
                working_dir, 
                requested_file_name
            )
```

I think all the code is duplicated in here. Maybe, we can extract it to a helper function.

Reply:
I agree but as commented before I was planning to finish with the four end-points to migrate all of them to make use of [program repository](https://github.com/Qiskit/qiskit-serverless/blob/main/gateway/api/repositories/programs.py).